### ISSUE TITLE
loss calculation

### ISSUE THREAD

**Issue by merv22** [senya-ashukha/quantile-regression-dqn-pytorch#2]
Hello.

I have noticed that when loss is calculated there is detach() statement on diff:

`loss = huber(diff) * (tau - (diff.detach() < 0).float()).abs()`

could you please explain the reason?

Thanks a lot.

**Comment by senya-ashukha** [comment]
Hi,

It is just funny way to write down a loss function that penalizes overestimation/underestimation errors with different weights, via the indicator function `loss(x) = |u| (τ if u >= 0 else 1-τ) = u (τ - Indicator(u<0))`. If  `u >= 0 -> Indicator(u<0) = 0 -> loss = |u| * τ`, and `if u < 0 ->  Indicator(u<0) = 1 -> -|u| (τ - 1) -> |u| * (1-τ)`.

Feel free to reach again! 
Arsenii

**Comment by merv22** [comment]
Hi,
Thank you very much. Now I see the asymmetric loss stated in the paper! This bring much clarification.
Could you please elaborate why this should be` (diff.detach() < 0).float()` and not `(diff < 0).float()`?
Why gradients should be discarded in this statement?

Thanks.
Behrouz

**Comment by senya-ashukha** [comment]
`(diff < 0).float()` is an indicator function 0/1, and its gradient is zero anywhere, so we dont need to compute it
There are actually ways to approximately differentiate through this the most popular is the straight-through estimator, but in this case, it is kind of not intended

**Comment by merv22** [comment]
My understanding of deep learning is as poor as my understanding of RL :)
Thanks again.
That was very helpful.
